pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,260 @@
+			logger.Debug("checking rgw health of object store %q", c.namespacedName.Name)
+			err := c.checkObjectStoreHealth()
+			if err != nil {
+				updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+				logger.Warningf("failed to check rgw health for object store %q. %v", c.namespacedName.Name, err)
+			}
+		}
+	}
+}
+
+func (c *bucketChecker) checkObjectStoreHealth() error {
+	/*
+		0. purge the s3 object by default
+		1. create an S3 user
+		2. always use the same user
+		3. if already exists just re-hydrate the s3 credentials
+		4. create a bucket with that user or use the existing one (always use the same bucket)
+		5. create a check file
+		6. get the hash of the file
+		7. PUT the file
+		8. GET the file
+		9. compare hashes
+		10. delete object on bucket
+		11. update CR health status check
+
+		Always keep the bucket and the user for the health check, just do PUT and GET because bucket creation is expensive
+	*/
+
+	var s3AccessKey string
+	var s3SecretKey string
+	s3endpoint := fmt.Sprintf("%s:%s", BuildDomainName(c.objContext.Name, c.namespacedName.Namespace), c.port)
+
+	// Generate unique user and bucket name
+	bucketName := genUniqueBucketName(c.objContext.UID)
+	userConfig := c.genUserConfig()
+
+	// Create S3 user
+	logger.Debugf("creating s3 user object %q for object store %q", userConfig.UserID, c.namespacedName.Name)
+	user, rgwerr, err := CreateUser(c.objContext, userConfig)
+	if err != nil {
+		if rgwerr == ErrorCodeFileExists {
+			user, _, err = GetUser(c.objContext, userConfig.UserID)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get details from ceph object user %q for object store %q", user.UserID, c.namespacedName.Name)
+			}
+		} else {
+			return errors.Wrapf(err, "failed to create object user %q. error code %d for object store %q", userConfig.UserID, rgwerr, c.namespacedName.Name)
+		}
+	}
+	// Set access and secret key
+	s3AccessKey = *user.AccessKey
+	s3SecretKey = *user.SecretKey
+
+	// Initiate s3 agent
+	logger.Debugf("initializing s3 connection for object store %q", c.namespacedName.Name)
+	s3client, err := NewS3Agent(s3AccessKey, s3SecretKey, s3endpoint)
+	if err != nil {
+		return errors.Wrap(err, "failed to initialize s3 connection")
+	}
+
+	// Force purge the s3 object before starting anything
+	err = cleanupObjectHealthCheck(s3client, c.objContext.UID)
+	if err != nil {
+		return errors.Wrapf(err, "failed to perform object cleanup for object store %q", c.namespacedName.Name)
+	}
+
+	// Bucket health test
+	err = c.testBucketHealth(s3client, bucketName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to run bucket health checks for object store %q", c.namespacedName.Name)
+	}
+
+	logger.Debug("successfully checked object store endpoint for object store %q", c.namespacedName.Name)
+
+	// Update the EndpointStatus in the CR to reflect the healthyness
+	// FIX ME ConditionConnected versus ConditionReady whether external or not
```

Comment:
This is my last issue to fix

Reply:
Why do we need a different status for external or converged? It's the same health check. What about a more specific condition for health checks such as `ConditionHealthy`?